yarn-project/txe/src/index.ts

```diff
@@ -26,60 +26,65 @@ type TXEForeignCallInput = {
+        [artifactPath] = (await readdir(targetPath)).filter(file => file.endsWith(`-${contractName}.json`));
+      }
+    }
+    this.logger.debug(`Loading compiled artifact ${artifactPath}`);
+    const artifact = loadContractArtifact(JSON.parse(await readFile(artifactPath, 'utf-8')));
+    inputs.splice(0, 2, artifact);
+  }
+
+  // eslint-disable-next-line camelcase
+  async resolve_foreign_call(callData: TXEForeignCallInput): Promise<ForeignCallResult> {
+    const { session_id: sessionId, function: functionName, inputs } = callData;
     this.logger.debug(`Calling ${functionName} on session ${sessionId}`);
 
     if (!TXESessions.has(sessionId) && functionName != 'reset') {
       this.logger.info(`Creating new session ${sessionId}`);
       TXESessions.set(sessionId, await TXEService.init(this.logger));
     }
 
-    if (functionName === 'reset') {
-      TXESessions.delete(sessionId) &&
-        this.logger.info(`Called reset on session ${sessionId}, yeeting it out of existence`);
-      return toForeignCallResult([]);
-    } else {
-      if (functionName === 'deploy') {
-        const pathStr = fromArray(inputs[0] as ForeignCallArray)
-          .map(char => String.fromCharCode(char.toNumber()))
-          .join('');
-        const contractName = fromArray(inputs[1] as ForeignCallArray)
-          .map(char => String.fromCharCode(char.toNumber()))
-          .join('');
-        let artifactPath = '';
-        // Is not the same contract we're testing
-        if (!pathStr) {
-          // eslint-disable-next-line camelcase
-          artifactPath = join(root_path, './target', `${package_name}-${contractName}.json`);
-        } else {
-          // Workspace
-          if (pathStr.includes('@')) {
-            const [workspace, pkg] = pathStr.split('@');
-            const targetPath = join(root_path, workspace, './target');
-            this.logger.debug(`Looking for compiled artifact in workspace ${targetPath}`);
-            artifactPath = join(targetPath, `${pkg}-${contractName}.json`);
-          } else {
-            // Individual contract
-            const targetPath = join(root_path, pathStr, './target');
-            this.logger.debug(`Looking for compiled artifact in ${targetPath}`);
-            [artifactPath] = (await readdir(targetPath)).filter(file => file.endsWith(`-${contractName}.json`));
-          }
-        }
-        this.logger.debug(`Loading compiled artifact ${artifactPath}`);
-        const artifact = loadContractArtifact(JSON.parse(await readFile(artifactPath, 'utf-8')));
-        inputs.splice(0, 2, artifact);
+    switch (functionName) {
+      case 'reset': {
+        TXESessions.delete(sessionId) &&
+          this.logger.info(`Called reset on session ${sessionId}, yeeting it out of existence`);
+        return toForeignCallResult([]);
+      }
+      case 'deploy': {
+        // Modify inputs and fall through
+        await this.#processDeployInputs(callData);
```

Comment:
Oh wow yeah the fall through thing was definitely non obvious, 100% this would've been broken in a refactor that e.g. added a third case.